src/tribol/search/InterfacePairFinder.cpp

```diff
@@ -1,643 +1,873 @@
+    bool is_symm = m_coupling_scheme->getMeshId1() == m_coupling_scheme->getMeshId2();
+    if (is_symm)
+    {
+        // account for symmetry
+        numPairs = mesh1NumElems * (mesh1NumElems + 1) / 2;
+    }
+    ArrayT<bool> contactArray(numPairs, numPairs, m_coupling_scheme->getAllocatorId());
+    bool* inContact = contactArray.data();
 
-   /*! Returns the number of vertices in the mesh */
-   integer numVerts() const { return m_vertSet.size(); }
+    // Allocate memory for a counter
+    ArrayT<int> countArray(1, 1, m_coupling_scheme->getAllocatorId());
+    int* pCount = countArray.data();
 
-   /*! Returns the number of elements in the mesh */
-   integer numElems() const { return m_elemSet.size(); }
+    ContactMode cmode = m_coupling_scheme->getContactMode();
+
+    try
+    {
+      forAllExec(m_coupling_scheme->getExecutionMode(), numPairs,
+        [mesh1NumElems, mesh2NumElems, is_symm, inContact, mesh1, mesh2, cmode, 
+          pCount] TRIBOL_HOST_DEVICE (IndexT i)
+        {
+          IndexT fromIdx = i / mesh1NumElems;
+          IndexT toIdx = i % mesh2NumElems;
+          if (is_symm)
+          {
+            IndexT row = algorithm::symmMatrixRow(i, mesh1NumElems);
+            IndexT offset = row * (row + 1) / 2;
+            fromIdx = row;
+            toIdx = i - offset;
+          }
+          inContact[i] = geomFilter( fromIdx, toIdx, 
+                                     mesh1, mesh2,
+                                     cmode );
+          RAJA::atomicAdd<RAJA::auto_atomic>(pCount, static_cast<int>(inContact[i]));
+      });
+    }  // End of profiling block
+    catch(const std::exception& e)
+    {
+      std::cerr << e.what() << std::endl;
+    }
+    catch(...)
+    {
+      std::cerr << "unknown exception encountered during findInterfacePairs/geomFilter" << std::endl;
+    }
+    
+    ArrayT<int, 1, MemorySpace::Host> countArray_host(countArray);
+    SLIC_INFO("Found " << countArray_host[0] << " pairs in contact" );
+
+    auto& contactPairs = m_coupling_scheme->getInterfacePairs();
+    contactPairs.resize(countArray_host[0]);
+
+    countArray.fill(0);
+    auto pairs_view = m_coupling_scheme->getInterfacePairsView();
+    forAllExec(m_coupling_scheme->getExecutionMode(), numPairs,
+      [inContact, pCount, pairs_view, mesh1NumElems, mesh2NumElems, is_symm] 
+        TRIBOL_HOST_DEVICE (IndexT i)
+      {
+        // Filtering removed this case
+        if (!inContact[i])
+        {
+          return;
+        }
+        
+        IndexT fromIdx = i / mesh1NumElems;
+        IndexT toIdx = i % mesh2NumElems;
+        if (is_symm)
+        {
+          IndexT row = algorithm::symmMatrixRow(i, mesh1NumElems);
+          IndexT offset = row * (row + 1) / 2;
+          fromIdx = row;
+          toIdx = i - offset;
+        }
+
+        // get unique index for the array
+        auto idx = RAJA::atomicInc<RAJA::auto_atomic>(pCount);
+
+        pairs_view[idx] = InterfacePair(fromIdx, toIdx, true);
```

Comment:
So was the first kernal (above) simply marking which pairs are proximate based on geomFilter(), and here we are going and actually adding them to the InterfacePairs object on the coupling scheme? I'm assuming that the pairs_view is not const since we are adding to it?